### PR TITLE
Cache the backend's subscriber dimensions and evaluate rules against them

### PR DESCRIPTION
**Pull request by tonidero**
### Description

`subscriber` responses now may come with a new `dimensions` property including information used for rule evaluation. This PR reads those and cached them separately (reasoning is that these values might not be provided by fortress or other situations, and we would like to keep the latest valid value)

This PR also propagates this new field's properties into the rules engine as root properties. If there are any conflicts in names with one of the other SDK properties, the rule evaluation will fail.

---

> [!NOTE]
> **Medium Risk**
> Touches local rule evaluation scope and persistent cache behavior for customer info; wrong sticky-cache or null handling could change audience outcomes, though conflicts fail closed and tests cover cache edge cases.
> 
> **Overview**
> Adds support for the optional **`dimensions`** object on customer info responses so audience/checkpoint rules can read backend-supplied fields as **root-level** variables (separate from the reserved `backend.*` predicate hashes).
> 
> **Caching:** `DeviceCache` persists dimensions in their own per-user key when fresh network customer info includes a non-empty `dimensions` object; missing/empty payloads and re-caching cache-loaded customer info do **not** overwrite the last stored value. User cache clears remove this entry too.
> 
> **Rules engine:** New `SubscriberDimensionsProvider` parses the cached JSON into `RulesDimensionValue`s and is registered in `LocalRulesEvaluator`. Name collisions with other SDK dimensions still fail the whole snapshot.
> 
> **Null semantics:** Introduces `RulesDimensionValue.NullValue` and shared `JsonElement.asRulesDimensionValue()` so explicit JSON `null` is preserved for rules (falsy, `var` resolves) rather than treated as unreadable; backend predicate metadata parsing uses the same helper.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8f97bcbaaaffb06d1ff67c698b921cec17e12146. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt [hunk 1]
```diff
@@ -424,6 +439,36 @@ public open class DeviceCache(
...
 
+    // region subscriber dimensions
+
+    @VisibleForTesting
+    internal fun subscriberDimensionsCacheKey(appUserID: String) = "$subscriberDimensionsCacheBaseKey.$appUserID"
+
+    @Synchronized
+    internal fun cacheSubscriberDimensions(appUserID: String, dimensionsJson: String) {
+        preferences.edit()
+            .putString(subscriberDimensionsCacheKey(appUserID), dimensionsJson)
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt:450]: I honestly thought whether we should store this in a file... But I don't expect this to grow too big... so shared preferences was easier.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt [hunk 2]
```diff
@@ -241,6 +247,15 @@ public open class DeviceCache(
 
     @Synchronized
     internal fun cacheCustomerInfo(appUserID: String, info: CustomerInfo) {
+        // The dimensions sibling of `subscriber` is cached on its own because not every response carries it:
+        // the last value received stays authoritative until a response replaces it. Absent or empty keeps the
+        // previous value, and a cache-loaded info never writes so a stale copy inside the blob below can't
+        // resurrect an older value.
```
Comment (rickvdl) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt:253]: So I don't think we'd have a way to 'clear' / 'reset' a value now? Let's say the backend returns a timestamp like `firstXAt`, whenever the SDK returns null / the field is omitted it will use the previous cached value

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import kotlinx.serialization.json.JsonObject
+import java.util.Date
+
+/**
+ * The dimensions the backend last delivered alongside the subscriber, cached because not every response carries
+ * them, exposed as root-level names. Distinct from the reserved nested `backend` root, which holds the backend's
+ * pre-evaluated predicate results for one evaluation.
+ *
+ * The names are the backend's to choose, and the root-name contract applies to it like any other source: one
+ * that collides with an SDK-provided dimension fails the snapshot. A value no rule could read is dropped here,
```
Comment (rickvdl) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt]: Hm I'm not too sure about this, if someone accidentally added a new field in the backend it would break checkpoints for everyone?

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt [hunk 4]
```diff
@@ -403,6 +404,9 @@ internal class PurchasesFactory(
                             Purchases.sharedInstance.purchasesOrchestrator.awaitCustomerInfo(appUserID)
                         },
                     ),
+                    SubscriberDimensionsProvider {
+                        cache.getCachedSubscriberDimensionsJson(identityManager.currentAppUserID)
```
Comment (rickvdl) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:409]: Should we have an integration test covering the user ID switch?

**Review by rickvdl — CHANGES_REQUESTED** [review]

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt (hunk 2 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt:253]: Yes that's correct... That was a bit of a discussion I had with myself/the agent 😅 I thought that it was fine, since the values would get updated the next time we get values + if the rules are updated, outdated values would be ignored right?

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt (hunk 3 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt]: Well, I think this is something I would like to know beforehand TBH... I don't think we should allow the backend to send the same properties that the SDK is setting.... We could establish some priorities, but that would make possible issues more "easy to miss". FWIW, these properties are ones that the backend sends, not something that a developer can add by themselves (which are namespaced under either `subscriber_attributes` or `custom`). But well, let's get @limdauto thoughts just in case 🙏

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt (hunk 4 above)
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesFactory.kt:409]: Right... I have been postponing integration tests until I can test this fully 😅 but yeah, I would like to have a PR after testing it with a bunch of integration tests, so I will include this 👍

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt (hunk 2 above)
Comment (rickvdl) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt:253]: Well yes, but I think there is a case where the updated value = the value was removed? That case can't be represented currently.

Comment (rickvdl) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/caching/DeviceCache.kt:253]: As discussed over Slack I think I misunderstood, as long as the object is returned it overrides the entire cached object, which seems sufficient :)

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/RulesDimensionValue.kt [hunk 5]
```diff
@@ -30,6 +30,15 @@ public sealed class RulesDimensionValue {
...
     public data class DateValue(val value: Date) : RulesDimensionValue()
 
+    /**
+     * An explicit null a source states for a name it does have, as opposed to leaving the name out.
+     *
+     * Reaches the engine as JSON `null`: `var` resolves it instead of failing as an unresolved variable, `missing`
+     * still reports it, and it is falsy. For a source whose values are its own to state — the backend's — rather
+     * than one the SDK observes, where an unavailable value is omitted.
+     */
+    public object NullValue : RulesDimensionValue()
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/RulesDimensionValue.kt:40]: @rickvdl I added this to be able to pass explicit null values from the backend to the rules engine, which until now were filtered out.

Comment (rickvdl) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/RulesDimensionValue.kt:40]: I think this makes sense, thanks for letting me know! Will do the same on iOS

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt (hunk 3 above)
Comment (rickvdl) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/localrules/SubscriberDimensionsProvider.kt]: Alright, let's discuss what we want to do here in a follow up, for now I don't think it's blocking indeed

**Review by rickvdl — APPROVED** [review]
